invenio_rdm_records/services/services.py

```diff
@@ -323,92 +323,97 @@ def purge_record(self, identity, id_, uow=None):
-        return super().search(identity, params, search_preference, expand, **kwargs)
+        return super().search(
+            identity,
+            params,
+            search_preference,
+            expand,
+            extra_filter=search_filter,
+            **kwargs,
+        )
 
-    def search_marked_for_purge(
-        self, identity, params=None, search_preference=None, expand=False, **kwargs
+    def search_all(
+        self,
+        identity,
+        params=None,
+        search_preference=None,
+        expand=False,
+        extra_filter=None,
+        **kwargs,
     ):
-        """Search for records marked for purge, matching the querystring."""
-        status = RecordDeletionStatusEnum.MARKED.value
-        kwargs = self._add_deletion_filter(kwargs, status)
-        return super().search(identity, params, search_preference, expand, **kwargs)
+        """Search for all (published and deleted) records matching the querystring."""
+        self.require_permission(identity, "search_all")
+
+        # exclude drafts filter (drafts have no deletion status)
+        search_filter = dsl.Q(
+            "terms", **{"deletion_status": [v.value for v in RecordDeletionStatusEnum]}
+        )
+        if extra_filter:
+            search_filter &= extra_filter
+        search_result = self._search(
+            "search_all",
+            identity,
+            params,
+            search_preference,
+            record_cls=self.draft_cls,
+            search_opts=self.config.search_all,
+            extra_filter=search_filter,
+            permission_action="read",  # TODO this probably should be read_deleted
+            **kwargs,
+        ).execute()
+
+        return self.result_list(
+            self,
+            identity,
+            search_result,
+            params,
+            links_tpl=LinksTemplate(self.config.links_search, context={"args": params}),
+            links_item_tpl=self.links_item_tpl,
+            expandable_fields=self.expandable_fields,
+            expand=expand,
+        )
 
     def search_drafts(
-        self, identity, params=None, search_preference=None, expand=False, **kwargs
+        self,
+        identity,
+        params=None,
+        search_preference=None,
+        expand=False,
+        extra_filter=None,
+        **kwargs,
     ):
         """Search for drafts that have not been marked as deleted."""
         # because drafts don't have a 'deletion_status', a simple positive filter
         # won't work in cases where records and drafts are mixed...
         # instead, we exclude all the 'wrong' values here
-        status = RecordDeletionStatusEnum.PUBLISHED.value
-        kwargs = self._add_deletion_filter(kwargs, status, positive_filter=False)
+        # TODO this filter can be removed if we cleanup "orphaned" drafts after deletion
+        deletion_filter = dsl.query.Bool(
+            "must_not", **{"exists": {"field": "deletion_status"}}
```

Comment:
Simplified filter from the removed `_add_deletion_filter`

Reply:
from my understanding, it's expected that:
* drafts don't have the `deletion_status` field
* (proper) records do have this field

which entails that this query filter would remove all published records from the search results, and only drafts would be considered.
is that intended? (from the method's name I'd assume, yes :sweat_smile: )

of course, I might have missed something and the above assumptions might not even hold true in practice

Reply:
I have tested the query it seems to have the same outcome as the previous one